daal4py/sklearn/neighbors/knn.py

```diff
@@ -0,0 +1,525 @@
+            Test samples.
+
+        Returns
+        -------
+        y : ndarray of shape (n_queries,) or (n_queries, n_outputs)
+            Class labels for each data sample.
+        """
+        X = check_array(X, accept_sparse='csr')
+
+        try:
+            fptype = getFPType(X)
+        except ValueError:
+            fptype = None
+
+        if self.weights in ['uniform', 'distance'] and self.algorithm in ['brute', 'kd_tree'] \
+        and (self.metric == 'minkowski' and self.p == 2 or self.metric == 'euclidean') \
+        and self._y.ndim == 1 and fptype is not None:
+
+            n_classes = len(self.classes_)
+
+            if self.algorithm == 'brute':
+                knn_classification_training = d4p.bf_knn_classification_training
+                knn_classification_prediction = d4p.bf_knn_classification_prediction
+            else:
+                knn_classification_training = d4p.kdtree_knn_classification_training
+                knn_classification_prediction = d4p.kdtree_knn_classification_prediction
+
+            alg_params = {
+                'fptype': fptype,
+                'method': 'defaultDense',
+                'k': self.n_neighbors,
+                'nClasses': n_classes,
+                'voteWeights': 'voteUniform' if self.weights == 'uniform' else 'voteDistance',
+                'resultsToEvaluate': 'computeClassLabels'
+            }
+
+            training_alg = knn_classification_training(**alg_params)
+
+            fit_X = d4p.get_data(self._fit_X)
+            _y = d4p.get_data(self._y)
+            _y = _y.reshape(_y.shape[0], 1)
+            training_result = training_alg.compute(fit_X, _y)
+
+            prediction_alg = knn_classification_prediction(**alg_params)
+
+            X = d4p.get_data(X)
+            prediction_result = prediction_alg.compute(X, training_result.model)
+            return prediction_result.prediction.ravel().astype(self._y.dtype)
+
+        else:
+            neigh_dist, neigh_ind = self.kneighbors(X)
+            classes_ = self.classes_
+            _y = self._y
+            if not self.outputs_2d_:
+                _y = self._y.reshape((-1, 1))
+                classes_ = [self.classes_]
+
+            n_outputs = len(classes_)
+            n_queries = _num_samples(X)
+            weights = _get_weights(neigh_dist, self.weights)
+
+            y_pred = np.empty((n_queries, n_outputs), dtype=classes_[0].dtype)
+            for k, classes_k in enumerate(classes_):
+                if weights is None:
+                    mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
+                else:
+                    mode, _ = weighted_mode(_y[neigh_ind, k], weights, axis=1)
+
+                mode = np.asarray(mode.ravel(), dtype=np.intp)
+                y_pred[:, k] = classes_k.take(mode)
+
+            if not self.outputs_2d_:
+                y_pred = y_pred.ravel()
+
+            return y_pred
+
+    def predict_proba(self, X):
```

Comment:
If we don't patch this function, why copy the content sklearn?

Reply:
Deleted